### PR TITLE
Implemented Array.sort and Array.reverse

### PR DESCRIPTION
**Pull request by bluntworks**
I gave this a go, I think I've caught all the subtleties, though let me know if I missed something.

### CODE REVIEW

File: array-sort.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    var sorted = unpacked
+            .map(function(it) { return it.val })
+            .sort(compare)
+
+    var packed = repack(sorted, unpacked)
+
+    //fake diff - for perf
+    //adiff on 10k items === ~3200ms
+    //fake on 10k items === ~110ms
+    var changes = [ [ 0, packed.length ].concat(packed) ]
```
Comment (Raynos) [inline array-sort.js]: Having the diff say "oh btw, everything was removed and everything was added" seems sub optimum.

The real challenge of this implementation was to do a complete immutable implementation of `sort()` in userland that build up the diff as it went along.

**Comment by Raynos** [comment]
lgtm. I had hoped a more clever _diff computation but that might be too hard.

**Comment by bluntworks** [comment]
How about either passing a boolean or a separate method ? to allow using adiff.diff if performance isn't an issue e.g. a relatively short array.

Also the bottleneck in [adiff is the score](https://github.com/dominictarr/adiff/blob/master/index.js#L28) function not sure if there is some optimization that could be done there.

**Comment by Raynos** [comment]
@bluntworks the optimization is to do it one go.

Instead of doing `var new = old.slice().sort(); adiff(old, new);` you want to re-implement sort in javascript so that it returns a diff and a new array.

Computing a delta between the two arrays is expensive, but building up the delta whilst you are sorting is a lot cheaper.

**Comment by bluntworks** [comment]
ahh ok interesting will def give it a go when I get some time to spend on it. 

cheers

**Comment by Raynos** [comment]
@mmckegg I'm thinking of merging this, would you like to do a quick review before i do ?

**Comment by mmckegg** [comment]
looks pretty good to me (have been following along).

I agree that a fast **and accurate** diff would be preferable, but yeah, no reason why the diff can't improve later. Have been finding the slowness of adiff on arr.set quite noticeable for my application and have moved most stuff where I can to standard `Observ([])` to get around it (was mostly using it as a backing for grid, where lcs diffing isn't actually needed). 

I think @bluntworks has made the correct tradeoff for the time being.

**Comment by Raynos** [comment]
Published v3.2.0

Added @bluntworks as a collaborator (you have commit access)

Added CONTRIBUTION.md to communicate contribution guidelines for collaborators

**Pull request merged**